### PR TITLE
Reworking how buffer mapping is used in prep for unmappable memory.

### PR DESCRIPTION
**Pull request by benvanik**
Metal, WebGPU, and remoting HAL backends map memory in special ways we haven't encountered yet in CUDA/Vulkan/local CPU execution. Many of the tools and samples we have were written assuming that mapping would always succeed, also requiring that the compiler emit buffers that were mappable. This set of changes is designed to make mapping needed less frequently, allow HAL backends to override transfer behavior, and add utilities for users to make it easier to work with data.

There's still some places that are doing extremely shady things (like the e2e matmul tests) but most are now in a form that is compatible with unmappable memory.

### CODE REVIEW

File: iree/hal/command_buffer.h [hunk 1]
```diff
@@ -476,6 +476,68 @@ IREE_API_EXPORT iree_status_t iree_hal_command_buffer_dispatch_indirect(
...
+} iree_hal_transfer_command_type_t;
+
+// Represents a single transfer command within a batch of commands.
+typedef struct iree_hal_transfer_command_t {
+  // The type of the command selecting which of the payload data is used.
+  iree_hal_transfer_command_type_t type;
+  union {
+    // IREE_HAL_TRANSFER_COMMAND_TYPE_FILL
+    struct {
+      iree_hal_buffer_t* target_buffer;
```
Comment (ScottTodd) [inline iree/hal/command_buffer.h:500]: struct -> union -> struct
eek - I'm wondering if there's a cleaner / easier to use way to represent this

File: iree/modules/check/module.cc [hunk 2]
```diff
@@ -220,6 +220,9 @@ class CheckModuleState final {
     iree_hal_element_type_t rhs_element_type =
         iree_hal_buffer_view_element_type(rhs);
 
+    // HACK: this is all broken and will leak. Let's kill this entire module
```
Comment (ScottTodd) [inline iree/modules/check/module.cc:223]: 😆 yep

File: iree/hal/command_buffer.h (hunk 1 above)
Comment (benvanik) [reply iree/hal/command_buffer.h:500]: This _is_ the easy way to do this kind of thing :) Anything else is going to be the same but with a lot more boilerplate and ambiguity: anonymous unions and structs are really useful ways of encapsulating data structures like this where there are zero valid uses outside of being a member of a particular structure.

File: iree/hal/buffer_view.h [hunk 3]
```diff
@@ -251,6 +254,39 @@ IREE_API_EXPORT iree_status_t iree_hal_buffer_view_wrap_or_clone_heap_buffer(
...
+// Buffers allocated like this do not need the IREE_HAL_BUFFER_USAGE_MAPPING bit
+// set; it will be added automatically if the allocator needs it and otherwise
+// the memory can remain unmappable (and thus fully device isolated).
+//
+// As this _may_ require allocation of the entire buffer content in host memory
+// it is always preferable to stage and issue copy commands via the device
+// queue. Even better is to do all generation on-device via dispatches without
+// the need to ever transfer. Usage of this method should be limited to times
+// where device-side generation isn't possible or memory consumption is not a
+// concern.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/buffer_view.h:275]: I'd thought about separating things too but was procrastinating - I'll move the helpers to buffer_view_util. Depending on the application this method is fine and saves a lot of easy-to-mess-up boilerplate - testonly stuff should either be really testonly (touching the test infra like including gtest.h, etc), unsafe/unstable/etc (don't want production users to accidentally use things), or unconditionally increase binary size/deps (anything that is LTO strippable is generally fine, but anything that incurs allocations/code size isn't).

File: iree/tools/iree-e2e-matmul-test.c [hunk 4]
```diff
@@ -478,21 +478,23 @@ static iree_status_t allocate_buffer_like(iree_hal_allocator_t* hal_allocator,
...
-      IREE_HAL_BUFFER_USAGE_ALL, dst);
+      IREE_HAL_BUFFER_USAGE_ALL, iree_const_byte_span_empty(), dst);
 }
 
 // Performs a deep copy of |src| into |dst|. Takes care of allocating |dst|.
 static iree_status_t copy_buffer(iree_hal_allocator_t* hal_allocator,
                                  iree_hal_buffer_view_t* src,
                                  iree_hal_buffer_view_t** dst) {
+  // TODO(benvanik): change this to use iree_hal_buffer_copy_data. Or something.
+  // I can't understand what all this code is doing.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/tools/iree-e2e-matmul-test.c:490]: scary - and won't generalize - my hope with some of the helpers I'm adding here is that stuff like this won't need to be written (or will be much harder to write than adapting your style to be able to use a helper that does things for you). But until forced I don't want to change any of this :P

[the remark this answers is not in the document]
Comment (benvanik) [reply iree/tools/iree-e2e-matmul-test.c:490]: (with the new hal.executable.source ops this whole bit of infra could change to be much simpler, but... until forced lol)

File: iree/hal/allocator.h [hunk 5]
```diff
@@ -121,6 +121,9 @@ iree_hal_allocator_query_buffer_compatibility(
     iree_hal_buffer_usage_t intended_usage, iree_device_size_t allocation_size);
 
 // Allocates a buffer from the allocator.
+// If |initial_data| is provided then the bytes will be copied into the device
+// buffer. To avoid the copy when constant data is used prefer
+// iree_hal_allocator_wrap_buffer when available.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/allocator.h:126]: nah - those are actually harder to understand: what we're talking about here is the **initial data**, not **the helper struct combining a pointer and a size** - "if initial data is provided then the bytes..." vs "if the helper struct combining a pointer and a size has a non-NULL and non-zero value pair then the bytes...".

**Review by ScottTodd — APPROVED** [review]
mostly trusting my initial review :P
I scanned through `iree/hal/buffer_view_util.h` and the updates to sample and CTS files

**Pull request merged**